### DISCUSSION TITLE
XML support?

### DISCUSSION THREAD

**Discussion by ethindp** [stephenberry/glaze#1319] [Q&A]
Would it be possible to add XML support? I know that XML is quite a bit more complicated with attributes as well as elements, but I figure that we could probably extend the current meta system to allow you to tell Glaze which are attributes and which are sub-elements. An alternative to this would be to take into consideration that attributes are strings in XML, so we could make our own set of types, and then employ the following rules:

1. If the type of a field in a struct is one of our types (i.e., `glz::date_time`, `glz::xml_string`, etc.), and The value read in matches the regular expression for that type, then:
2. The field is an attribute, so treat it as one.
3. Otherwise, if the type is NOT one of those specific types, it is an element.

This would solve the problem of purely textual elements, for example, like:

```xml
<a attr1="2" attr2="true">
<a>true</a>
</a>
```

Could be theoretically defined (using this alternative model) as:

```cpp
struct a {
  glz::xml::integer attr1; // XML integer as defined by the XSD spec
  glz::xml::boolean attr2; // XML boolean as defined by the XSD spec
  std::string a; // Element, not attribute
};
```

Would either of these models be feasible/doable? I imagine the first would probably be easier; the second may require the XML XSD types to be actual classes with things like `operator std::uint32_t` overloads so that they satisfy `std::convertible_to`, perhaps. If this isn't on the roadmap that's alright, but I thought I would ask.

**Comment by stephenberry** [comment]
I don't have the capacity right now to add XML support. Although I am working on improving the library so that it is easier to add new formats.

If you wanted to implement XML support I could help orient you with the architecture. But, I feel like this would be a large effort.

What is your motivation for parsing XML directly into C++ objects? Are you doing web development or do you have XML input files?

**Comment by ethindp** [comment]
I do have a project that sadly must read XML, yes, but it's not a project that I need to complete on a deadline or anything, so I've been looking around and trying different solutions to see what works best. I thought of writing an XSD to C++ converter but I wondered if maybe something like this library could do it instead of me having to write a converter.

**Comment by stephenberry** [comment]
I would recommend looking for another C++ library that can handle XML for you.

In the long term we do plan to add more formats to Glaze, but XML is a complex spec that would take a while to implement.

But, thanks for inquiring.

**Comment by stephenberry** [comment]
I am going to move this to the discussions page for future reference.
